app/models/server.rb

```diff
@@ -143,6 +143,18 @@ def reset_unhealthy_counter
     end
   end
 
+  # Validates uniqueness of a new server object, and attempts to save it. Raises a RecordNotSaved exception on errors,
+  # otherwise returns the newly created object.
+  def self.create!(attributes = {})
+    Server.all.each_slice(1000) do |server_list|
+      server_list.each do |server|
+        raise RecordNotSaved.new('Server url is not unique', self) if server.url.eql?(attributes[:url])
+        raise RecordNotSaved.new('Server id is not unique', self) if server.secret.eql?(attributes[:secret])
+      end
+    end
```

Comment:
There's no need to scan through all of the servers - you can simply check for the existence of the server id in the 'servers' set. That said, there's still a race condition in that case.

To do this correctly, the `create!` method needs to add the server inside a redis transaction (redis.multi) with a `WATCH` on the 'servers' set so that the transaction will abort if there's any concurrent modifications that add or remove servers.

Reply:
Oh, I'm getting this mixed up with unique server ids. Yeah, solving the issue of unique server urls (I wouldn't bother checking secret) is gonna be a pain to do, if we wanted to support it we should be adding an extra set to contain servers urls that can be checked against.

The original design included no protection against duplicate servers.

Note that checking only in create is not sufficient, because the url can be updated.